x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/IndicesPermission.java

```diff
@@ -202,43 +229,83 @@ public Automaton allowedActionsMatcher(String index) {
+            boolean granted = false;
+            // true if ANY group, which contains certain ingest privileges, covers the given index AND the action is a mapping update for
+            // an index or an alias (but not for a data stream)
+            boolean bwcGrantMappingUpdate = false;
+            final List<Runnable> bwcDeprecationLogActions = new ArrayList<>();
+
             for (Group group : groups) {
-                // check for privilege granted directly on the requested index/alias
-                if (group.check(action, indexOrAlias) ||
-                    // check for privilege granted on parent data stream if a backing index
-                    (indexAbstraction != null && indexAbstraction.getType() == IndexAbstraction.Type.CONCRETE_INDEX &&
-                        indexAbstraction.getParentDataStream() != null &&
-                        group.check(action, indexAbstraction.getParentDataStream().getName()))) {
-                    granted = true;
-                    for (String index : concreteIndices) {
-                        Set<FieldPermissions> fieldPermissions = fieldPermissionsByIndex.computeIfAbsent(index, (k) -> new HashSet<>());
-                        fieldPermissionsByIndex.put(indexOrAlias, fieldPermissions);
-                        fieldPermissions.add(group.getFieldPermissions());
-                        DocumentLevelPermissions permissions =
-                                roleQueriesByIndex.computeIfAbsent(index, (k) -> new DocumentLevelPermissions());
-                        roleQueriesByIndex.putIfAbsent(indexOrAlias, permissions);
-                        if (group.hasQuery()) {
-                            permissions.addAll(group.getQuery());
-                        } else {
-                            // if more than one permission matches for a concrete index here and if
-                            // a single permission doesn't have a role query then DLS will not be
-                            // applied even when other permissions do have a role query
-                            permissions.setAllowAll(true);
+                // the group covers the given index OR the given index is a backing index and the group covers the parent data stream
+                final boolean indexCheck = group.checkIndex(indexOrAlias) ||
+                        (isBackingIndex && group.checkIndex(indexAbstraction.getParentDataStream().getName()));
+                if (indexCheck) {
+                    boolean actionCheck = group.checkAction(action);
+                    // mapping updates are allowed for certain privileges on indices and aliases (but not on data streams),
+                    // outside of the privilege definition
+                    boolean bwcMappingActionCheck = isMappingUpdateAction && false == isDataStream && false == isBackingIndex &&
+                            group.privilege().name().stream().anyMatch(privilegeNameSetBwcAllowMappingUpdate::contains);
```

Comment:
I don't like this minor optimisation because it makes it harder to reason about the meaning of the check.
There is some overlapping between the `actionCheck` and the `bwcMappingActionCheck` flags (they can both be true) because it is easier to express them.

Reply:
I do agree this by itself makes the code harder to read. Since I made a few comments around this code block, I think what I am really suggesting is to re-arrange in a broader context, e.g.:

```java
if (group.checkAction(action)) {
    granted = true;
    propagateDlsAndFlsPermissionsOvertheConcreteIndices(...);
} else if (isMappingUpdateAction && false == isDataStream && false == isBackingIndex &&
        group.privilege().name().stream().anyMatch(privilegeNameSetBwcAllowMappingUpdate::contains)) {
    bwcGrantMappingUpdate = true;
    propagateDlsAndFlsPermissionsOvertheConcreteIndices(...);
    populateDeprecationLogs(...);
}
```

Reply:
I understand now, but I don't see an improvement. `bwcGrantMappingUpdate` is updated to `true` in the if branch, but I slightly prefer it when its value is the expression of the if branch itself; I find it easier to follow like this.
Also, you need a similar flag for the `if (group.checkAction(action))` if branch, because deprecation logs must only be present if the action is exclusively granted by the deprecation behaviour.

Reply:
I don't quite follow. 

Once the logic is inside the `else if` branch, we can already be sure the action is granted by bwc and can go ahead and prepare the deprecation logs within the branch (i.e. `populateDeprecationLogs(...)`). With this approach, we can get rid of two variables `actionCheck` and `bwcMappingActionCheck`. We still need `granted` and `bwcGrantMappingUpdate`, but only one of them needs to be updated for each loop.

Reply:
Sorry, you're right, I got lost.
I'm tempted to not change the code as you're suggesting, because the goal to get rid of `actionCheck` and `bwcMappingActionCheck` is not worthwhile IMO. To me these look like different approaches with no substantial differences.